### PR TITLE
Added some preliminary docs

### PR DESCRIPTION
**Pull request by shuhaowu**
See: https://shopify.github.io/mybench

### CODE REVIEW

File: docs/detailed-design-doc.rst [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+requires the definition of the ``Event`` callback, a ``Config`` method
+returning the configurations of the workload (such as the event rate and number
+of parallel workers), and a ``NewContextData`` method that constructs a
+custom, thread-local data structure that could be useful for the workload to
+store thread-local state. Each custom benchmark is tied together with a
+``BenchmarkApp`` which constructs and owns the ``Benchmark`` object. Workloads
+defined by the user can be added to the ``Benchmark`` through the
+``BenchmarkApp`` object.
+
+Finally, the ``BenchmarkApp`` also handles other administrative duties common
```
Comment (ajmaidak) [inline docs/detailed-design-doc.rst:156]:
```suggestion

The ``BenchmarkApp`` also handles other administrative duties common
```

File: docs/detailed-design-doc.rst [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+stores the throughput and latency statistics during the benchmark run. Data
+held inside the ``OnlineHistogram`` objects are periodically read by the
+``DataLogger`` (see `High-performance data logging via double buffering`_).
+Once the data is read, the ``DataLogger`` aggregates the throughput and latency
+statistics across all ``BenchmarkWorker``\s and all ``Workload``\s into
+per-workload and overall throughput and latency statistics. These values are
+then written into a SQLite database.
+
+Users developing a custom benchmark do not need to be concerned with the
+internal details of mybench. To define a ``Workload``, the user simply needs to
```
Comment (ajmaidak) [inline docs/detailed-design-doc.rst:145]:
```suggestion
internal details of mybench. To define a ``Workload``, the user needs to
```

File: docs/detailed-design-doc.rst [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+"temporal-discretization" looper`_), and an ``OnlineHistogram`` object, which
+stores the throughput and latency statistics during the benchmark run. Data
+held inside the ``OnlineHistogram`` objects are periodically read by the
+``DataLogger`` (see `High-performance data logging via double buffering`_).
+Once the data is read, the ``DataLogger`` aggregates the throughput and latency
+statistics across all ``BenchmarkWorker``\s and all ``Workload``\s into
+per-workload and overall throughput and latency statistics. These values are
+then written into a SQLite database.
+
+Users developing a custom benchmark do not need to be concerned with the
```
Comment (ajmaidak) [inline docs/detailed-design-doc.rst:144]: Consider splitting this section into a usage section rather then leaving it in the "internals" section.

File: docs/detailed-design-doc.rst [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+``BenchmarkApp`` which constructs and owns the ``Benchmark`` object. Workloads
+defined by the user can be added to the ``Benchmark`` through the
+``BenchmarkApp`` object.
+
+Finally, the ``BenchmarkApp`` also handles other administrative duties common
+between all custom-defined user benchmarks, such as parsing the command line
+flags and setting an HTTP server where live throughput and latency statistics
+can be monitored.
+
+While mybench is primarily designed to benchmark MySQL, its architecture is
```
Comment (ajmaidak) [inline docs/detailed-design-doc.rst:161]: I suggest dropping this paragraph.

File: docs/detailed-design-doc.rst [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+flexible enough to benchmark arbitrary database and non-database systems.
+However, these are not uses cases that have been thoroughly tested for now.
+
+Rate control via "temporal-discretization" looper
+-------------------------------------------------
+
+A common approach to discover the limits of the database is by performing a
+sequence of benchmarks with increasing event rates. When the database is
+overloaded, the event rate will plateau and the latency will stagnate. To
```
Comment (ajmaidak) [inline docs/detailed-design-doc.rst:170]:
```suggestion
overloaded, the event rate will plateau or decline and the latency will increase. To
```

File: docs/detailed-design-doc.rst [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+   elapsed time. Number of events is calculated by summing the event count in
+   all ``OnlineHistogram`` instances. The elapsed time is calculated as the time
+   difference between two subsequent sampling of all ``OnlineHistogram``
+   instances. If the time required to read all ``OnlineHistogram`` instances is
+   a large fraction of the elapsed time between two subsequent
+   ``OnlineHistogram`` sampling, this can introduce significant bias in the
+   calculated throughput, as both numerator and the denominator will be affected
+   by the passage of time.
+
+Random data generation
```
Comment (ajmaidak) [inline docs/detailed-design-doc.rst:281]:
```suggestion
Data generation
```

File: docs/detailed-design-doc.rst [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+----------------------
+
+Database benchmarks require the generation of random data. Similar to other
```
Comment (ajmaidak) [inline docs/detailed-design-doc.rst:284]:
```suggestion
Database benchmarks require the generation of data. Similar to other
```

File: docs/introduction.rst [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  benchmarks.
+* **SQLite-based data logging**: mybench writes the throughput and latency time
+  series into a SQLite database. Multiple benchmark runs can share the same
+  database file. The standard and single-file approach to data logging allows
+  for better interoperability with other tools as well as for easier data
+  sharing.
+* **Post-processing scripts**: mybench includes a set of scripts that creates
+  standard visualizations with the data logged in the SQLite database file.
+  Multiple benchmark runs can be compared with these scripts.
+* **Random data generator with existing data sampling**: mybench includes a
```
Comment (ajmaidak) [inline docs/introduction.rst:33]:
```suggestion
* **Data generators with data sampling**: mybench includes a
```

**Review by ajmaidak — APPROVED** [review]
Excellent Documentation

File: docs/detailed-design-doc.rst (hunk 3 above)
Comment (shuhaowu) [reply docs/detailed-design-doc.rst:144]: I'm going to leave this here as I don't have an usage section.

File: docs/detailed-design-doc.rst [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+that already exists on the database such that the data can be used in the
+``WHERE`` clause of a ``SELECT`` statement. At this time, the data generator
+implemented within mybench are relatively primitive. For example, accurately
+sampling existing data can be memory and time intensive and mybench does not
+implement a fully-correct version of the existing data sampling algorithm at
+this time. Instead, data sampling is done in a best-effort manner to ensure
+the high performance and small resource footprint of mybench is not
+compromised.
+
+The random data generators implement in mybench uses the standard ``rand``
```
Comment (saunderst) [inline docs/detailed-design-doc.rst:294]:
```suggestion
The random data generators implemented in mybench use the standard ``rand``
```

**Review by saunderst — COMMENTED** [review]
This is excellent.  The "why mybench" proposition is very strongly presented.  I only have a few grammatical edits to offer so far.

**Review by saunderst — APPROVED** [review]

**Pull request merged**